### PR TITLE
Implement open-in-atom

### PR DESCRIPTION
**Pull request by tjkandala**
To the best of my knowledge, Atom core URI handlers support opening files only, not folders. See code comment for research sources. I hope I'm missing some information, but in case we can't find a way to open folders, this PR enables open-in-atom enabled on `CodeEditor` views only.

Part of https://github.com/sourcegraph/sourcegraph/issues/13270

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -3,11 +3,10 @@
...
   "activationEvents": [
     "*"
   ],
-  "wip": false,
+  "wip": true,
```
Comment (felixfbecker) [inline package.json:9]: **question**: Why WIP? What's missing?

File: src/open-in-atom.ts [hunk 2]
```diff
@@ -1,87 +1,60 @@
...
-        urlPattern: 'idea://open?file=%file&line=%line&column=%col',
-    },
-    sublime: {
-        urlPattern: 'subl://open?url=%file&line=%line&column=%col',
-    },
-    custom: {
-        urlPattern: '',
-    },
+interface Settings {
+    'openinatom.basePath'?: string
```
Comment (felixfbecker) [inline src/open-in-atom.ts:5]: **nitpick**: Settings are camelCase, so can we rename this to `openInAtom.basePath`?

File: src/open-in-atom.ts [hunk 3]
```diff
@@ -1,87 +1,60 @@
...
-    const editor: unknown = sourcegraph.configuration.get().value['openineditor.editor'] as unknown
-    const customUrlPattern: unknown = sourcegraph.configuration.get().value['openineditor.customUrlPattern'] as unknown
-    const learnMorePath = new URL('/extensions/sourcegraph/open-in-editor', sourcegraph.internal.sourcegraphURL.href).href
+    const basePath = sourcegraph.configuration.get<Settings>().value['openinatom.basePath']
+    const learnMorePath = new URL('/extensions/sourcegraph/open-in-atom', sourcegraph.internal.sourcegraphURL.href).href
 
     if (typeof basePath !== 'string') {
         throw new TypeError(
-            `Add \`openineditor.basePath\` to your user settings to open files in the editor. [Learn more](${learnMorePath})`
+            `Add \`openinatom.basePath\` to your user settings to open files in the editor. [Learn more](${learnMorePath})`
```
Comment (felixfbecker) [inline src/open-in-atom.ts:14]: **suggestion**: link "user settings" like in the other error below.

File: src/open-in-atom.ts [hunk 4]
```diff
@@ -1,87 +1,60 @@
...
-    let line = 1;
-    let column = 1;
+    const openUrl = new URL('atom://core/open/file?filename=' + absolutePath)
 
+    // This should always be true, since atom only supports file URIs
     if (sourcegraph.app.activeWindow?.activeViewComponent?.type === 'CodeEditor') {
         const selection = sourcegraph.app.activeWindow?.activeViewComponent?.selection
         if (selection) {
-            line = selection.start.line + 1
+            openUrl.search += `&line=${selection.start.line + 1}`
```
Comment (felixfbecker) [inline src/open-in-atom.ts:35]:
```suggestion
            openUrl.searchParams.set('line', (selection.start.line + 1).toString())
```

File: src/open-in-atom.ts [hunk 5]
```diff
@@ -1,87 +1,60 @@
...
             if (selection && selection.start.character !== 0) {
-                column = selection.start.character + 1
+                openUrl.search += `&column=${selection.start.character + 1}`
```
Comment (felixfbecker) [inline src/open-in-atom.ts:37]:
```suggestion
                openUrl.searchParams.set('column', (selection.start.character + 1).toString())
```

File: README.md [hunk 6]
```diff
@@ -1,21 +1,17 @@
...
-  - `idea` (JetBrains IntelliJ IDEA): `"openineditor.editor": "idea"`
-  - `sublime` (Sublime Text, requires a URL handler installed such as [this one for macOS](https://github.com/inopinatus/sublime_url))
-  - `custom` (requires also setting `openineditor.customUrlPattern`): `"openineditor.editor": "custom"`
-- `openineditor.basePath`: The absolute path on your computer where your git repositories live. This extension requires all git repos to be already cloned under this path with their original names. `"/Users/yourusername/src"` is a valid absolute path, while `"~/src"` is not.
-- `openineditor.customUrlPattern`: If you set `openineditor.editor` to `custom`, this must be set. Use placeholders `%file`, `%line`, and `%col` to mark where the file path, line number, and column number must be replaced. Example URL for IntelliJ IDEA: `idea://open?file=%file&line=%line&column=%col`
+- `openinatom.basePath`: The absolute path on your computer where your git repositories live. This extension requires all git repos to be already cloned under this path with their original names. `"/Users/yourusername/src"` is a valid absolute path, while `"~/src"` is not.
```
Comment (felixfbecker) [inline README.md:15]: **thought**: Maybe we could validate thing in the JSON schema with a `"pattern":  "[^~]+"`, so users don't need to find this hint in the README

File: README.md [hunk 7]
```diff
@@ -1,21 +1,17 @@
 # open-in-editor (Sourcegraph extension)
 
-Adds a button at the top of files in both Sourcegraph app and code hosts like GitHub (when the Sourcegraph browser extension is installed) that will open the current file in your editor of choice. 
+Adds a button at the top of files in both Sourcegraph app and code hosts like GitHub (when the Sourcegraph browser extension is installed) that will open the current file in Atom. 
 
-![Screenshot](https://user-images.githubusercontent.com/6304497/94799074-dcfd1d80-03e2-11eb-86fc-e55e03380154.png)
+![Screenshot](https://user-images.githubusercontent.com/37420160/96530024-856d0780-1254-11eb-968c-624aff1fd2e5.png)
```
Comment (felixfbecker) [inline README.md:5]: **tip**: You can define a light and dark version of the screenshot using `<picture>`: https://sourcegraph.com/github.com/sourcegraph/sourcegraph-code-stats-insights/-/blob/README.md#L6:1

File: src/open-in-atom.ts [hunk 8]
```diff
@@ -1,87 +1,60 @@
...
 export function activate(context: sourcegraph.ExtensionContext): void {
+    /**
+     * This doesn't seem to be documented, so here's where to find the URI pattern:
+     *
+     * Implementation ("core" URI handlers. Implements `openFile`, but not an equivalent `openFolder`):
+     * https://sourcegraph.com/github.com/atom/atom@320c879/-/blob/src/core-uri-handlers.js
+     *
+     * Tests ("core" host, only tests opening files):
+     * https://sourcegraph.com/github.com/atom/atom@320c879/-/blob/spec/main-process/atom-application.test.js#L677
+     */
```
Comment (felixfbecker) [inline src/open-in-atom.ts:54]: Thanks for documenting this in the code, this is very helpful 👏

**Review by felixfbecker — COMMENTED** [review]
Awesome! The most difficult done 🙂

File: README.md [hunk 9]
```diff
@@ -1,21 +1,17 @@
...
 
 ## Installing
 
-For Sourcegraph.com, it can be enabled here: https://sourcegraph.com/extensions/sourcegraph/open-in-editor
+For Sourcegraph.com, it can be enabled here: https://sourcegraph.com/extensions/sourcegraph/open-in-atom
 
 Otherwise, follow instructions here to publish it privately: https://docs.sourcegraph.com/extensions/authoring/publishing
```
[the remark this answers is not in the document]
Comment (tjkandala) [reply README.md:11]: Cool, I'll make these changes to open-in-editor as well

File: package.json (hunk 1 above)
Comment (tjkandala) [reply package.json:9]: Nothing I can think of, just didn't want to take it out of WIP before review

File: README.md (hunk 6 above)
Comment (tjkandala) [reply README.md:13]: Would `"pattern": "^\/+"` work better, since we want to assert that the path is absolute?

Comment (felixfbecker) [reply README.md:13]: Generally yeah we could define a pattern for that. Keep in mind though on Windows an absolute path generally looks like `c:\user data\felix`

Comment (tjkandala) [reply README.md:13]: Considering that, it may be easier to check for common invalid paths for now

File: README.md [hunk 10]
```diff
@@ -1,21 +1,15 @@
 # open-in-editor (Sourcegraph extension)
```
Comment (felixfbecker) [inline README.md:1]:
```suggestion
# Open in Atom Sourcegraph extension
```

File: README.md [hunk 11]
```diff
@@ -1,21 +1,15 @@
...
-
-## Installing
-
-For Sourcegraph.com, it can be enabled here: https://sourcegraph.com/extensions/sourcegraph/open-in-editor
-
-Otherwise, follow instructions here to publish it privately: https://docs.sourcegraph.com/extensions/authoring/publishing
+<picture>
+<source srcset="https://user-images.githubusercontent.com/37420160/96530024-856d0780-1254-11eb-968c-624aff1fd2e5.png" media="(prefers-color-scheme: dark)" />
+<source srcset="https://user-images.githubusercontent.com/37420160/96607497-6b6c0d00-12c6-11eb-921b-6ac6af8e90eb.png" media="(prefers-color-scheme: light)" />
+<img src="https://user-images.githubusercontent.com/37420160/96530024-856d0780-1254-11eb-968c-624aff1fd2e5.png" alt="Screenshot" />
```
Comment (felixfbecker) [inline README.md:8]: I'd use ligth as the default so GitHub will fall back to it
```suggestion
<img src="https://user-images.githubusercontent.com/37420160/96607497-6b6c0d00-12c6-11eb-921b-6ac6af8e90eb.png" alt="Screenshot" />
```

File: README.md [hunk 12]
```diff
@@ -1,21 +1,15 @@
...
-  - `idea` (JetBrains IntelliJ IDEA): `"openineditor.editor": "idea"`
-  - `sublime` (Sublime Text, requires a URL handler installed such as [this one for macOS](https://github.com/inopinatus/sublime_url))
-  - `custom` (requires also setting `openineditor.customUrlPattern`): `"openineditor.editor": "custom"`
-- `openineditor.basePath`: The absolute path on your computer where your git repositories live. This extension requires all git repos to be already cloned under this path with their original names. `"/Users/yourusername/src"` is a valid absolute path, while `"~/src"` is not.
-- `openineditor.customUrlPattern`: If you set `openineditor.editor` to `custom`, this must be set. Use placeholders `%file`, `%line`, and `%col` to mark where the file path, line number, and column number must be replaced. Example URL for IntelliJ IDEA: `idea://open?file=%file&line=%line&column=%col`
+- `openinatom.basePath`: The absolute path on your computer where your git repositories live. This extension requires all git repos to be already cloned under this path with their original names. `"/Users/yourusername/src"` is a valid absolute path, while `"~/src"` is not.
```
Comment (felixfbecker) [inline README.md:13]:
```suggestion
- `openInAtom.basePath`: The absolute path on your computer where your git repositories live. This extension requires all git repos to be already cloned under this path with their original names. `"/Users/yourusername/src"` is a valid absolute path, while `"~/src"` is not.
```

File: package.json [hunk 13]
```diff
@@ -88,5 +89,6 @@
...
...08 -26.30336,-0.6871 -83.03906,0.28014 -147.857502,23.11494 -149.509319,54.71024 -0.786577,15.04533 12.887399,29.634243 36.327769,41.741243' inkscape:connector-curvature='0' /%3E%3Cpath id='path8' style='fill:none;stroke:%23ffffff;stroke-width:15;stroke-linecap:round;stroke-miterlimit:4;stroke-dasharray:none;stroke-opacity:1' d='m 246.94025,-249.74892 c -16.87219,-14.033 -32.69622,-21.56881 -45.40971,-20.61296 -3.38495,0.25449 -6.5155,1.09073 -9.42465,2.57662 -31.03092,15.84954 -21.11773,97.82121 22.19078,183.025916 43.3085,85.20471863 103.57031,141.41684 134.60123,125.567287 29.84195,-15.242267 21.7875,-91.625309 -17.39278,-173.234763' inkscape:connector-curvature='0' /%3E%3C/g%3E%3C/g%3E%3Cg id='g6023'%3E%3Cpath id='path5965' style='fill:%23b3b3b3;fill-opacity:1;stroke:none;filter:url(%23filter6003)' d='m 292.90684,-107.76848 a 20.562887,20.612961 0 0 1 -41.12577,0 20.562887,20.612961 0 1 1 41.12577,0 z' inkscape:connector-curvature='0' /%3E%3Cpath inkscape:connector-curvature='0' d='m 290.90684,-109.76848 a 20.562887,20.612961 0 0 1 -41.12577,0 20.562887,20.612961 0 1 1 41.12577,0 z' style='fill:%23ffffff;fill-opacity:1;stroke:none' id='path10' /%3E%3C/g%3E%3C/g%3E%3C/svg%3E%0A"
```
Comment (felixfbecker) [inline package.json:93]: I used https://jakearchibald.github.io/svgomg/ to minify it before encoding (to 21.83% the size) 
```suggestion
    "icon": "data:image/svg+xml,%3Csvg viewBox='0 0 256 256' xmlns='http://www.w3.org/2000/svg' preserveAspectRatio='xMidYMid'%3E%3Cdefs/%3E%3Cdefs%3E%3ClinearGradient x1='50%25' y1='4%25' x2='50%25' y2='100%25' id='a'%3E%3Cstop stop-color='%23FDFDFD' offset='0%25'/%3E%3Cstop stop-color='%23CBCBCB' offset='100%25'/%3E%3C/linearGradient%3E%3ClinearGradient x1='50%25' y1='4%25' x2='50%25' y2='100%25' id='b'%3E%3Cstop stop-color='%23B2E198' offset='0%25'/%3E%3Cstop stop-color='%2304A171' offset='100%25'/%3E%3C/linearGradient%3E%3C/defs%3E%3Ccircle fill='url(%23a)' cx='128' cy='128' r='127.9'/%3E%3Ccircle fill='url(%23b)' cx='127.9' cy='128.1' r='116.9'/%3E%3Cpath d='M100 110l-4 6-3-11 12-1-5 6zm104 8c-5-4-10-8-18-11a172 172 0 00-64-12c17-20 32-30 41-30l5 1c5 3 6 11 4 23a4 4 0 008 2c4-20-2-29-7-32-3-2-6-3-10-3-14 1-34 16-52 39l-20 2-2-10c-2-13 0-23 5-25l3-1c4 0 11 3 18 9a4 4 0 005-7c-9-7-17-11-24-10l-5 1c-5 3-8 7-10 14v20l1 10-14 4c-18 7-22 16-22 22-1 6 2 14 18 23a4 4 0 004-8c-9-5-14-10-14-15 0-6 11-13 30-17l7 19c-7 13-12 25-14 36-2 8-3 15-2 21 1 7 4 11 8 14 2 2 5 3 9 3 6 0 13-3 23-10a4 4 0 10-5-7c-9 7-18 10-22 7-5-3-6-13-3-26l10-28 4 7c9 19 21 36 33 47 6 5 12 9 17 11 3 2 6 2 9 2l7-1c5-3 8-7 10-13l1-20c-2-15-8-33-16-51a4 4 0 00-8 3c8 18 14 35 15 49 1 13-1 22-5 24-5 3-14-1-24-11a163 163 0 01-38-56 211 211 0 0115-24 199 199 0 0168 12c12 5 19 11 19 17 0 5-6 9-16 13a4 4 0 003 8c17-6 21-15 21-21 1-5-2-10-6-14z' fill='%23F2F8F6'/%3E%3Cpath d='M119 129a9 9 0 1019 0 9 9 0 00-19 0' fill='%23F1F8F3'/%3E%3C/svg%3E"
```

File: package.json [hunk 14]
```diff
@@ -19,38 +18,40 @@
...
```
Comment (felixfbecker) [inline package.json:30]:
```suggestion
          "iconURL": "data:image/svg+xml,%3Csvg viewBox='0 0 256 256' xmlns='http://www.w3.org/2000/svg' preserveAspectRatio='xMidYMid'%3E%3Cdefs/%3E%3Cdefs%3E%3ClinearGradient x1='50%25' y1='4%25' x2='50%25' y2='100%25' id='a'%3E%3Cstop stop-color='%23FDFDFD' offset='0%25'/%3E%3Cstop stop-color='%23CBCBCB' offset='100%25'/%3E%3C/linearGradient%3E%3ClinearGradient x1='50%25' y1='4%25' x2='50%25' y2='100%25' id='b'%3E%3Cstop stop-color='%23B2E198' offset='0%25'/%3E%3Cstop stop-color='%2304A171' offset='100%25'/%3E%3C/linearGradient%3E%3C/defs%3E%3Ccircle fill='url(%23a)' cx='128' cy='128' r='127.9'/%3E%3Ccircle fill='url(%23b)' cx='127.9' cy='128.1' r='116.9'/%3E%3Cpath d='M100 110l-4 6-3-11 12-1-5 6zm104 8c-5-4-10-8-18-11a172 172 0 00-64-12c17-20 32-30 41-30l5 1c5 3 6 11 4 23a4 4 0 008 2c4-20-2-29-7-32-3-2-6-3-10-3-14 1-34 16-52 39l-20 2-2-10c-2-13 0-23 5-25l3-1c4 0 11 3 18 9a4 4 0 005-7c-9-7-17-11-24-10l-5 1c-5 3-8 7-10 14v20l1 10-14 4c-18 7-22 16-22 22-1 6 2 14 18 23a4 4 0 004-8c-9-5-14-10-14-15 0-6 11-13 30-17l7 19c-7 13-12 25-14 36-2 8-3 15-2 21 1 7 4 11 8 14 2 2 5 3 9 3 6 0 13-3 23-10a4 4 0 10-5-7c-9 7-18 10-22 7-5-3-6-13-3-26l10-28 4 7c9 19 21 36 33 47 6 5 12 9 17 11 3 2 6 2 9 2l7-1c5-3 8-7 10-13l1-20c-2-15-8-33-16-51a4 4 0 00-8 3c8 18 14 35 15 49 1 13-1 22-5 24-5 3-14-1-24-11a163 163 0 01-38-56 211 211 0 0115-24 199 199 0 0168 12c12 5 19 11 19 17 0 5-6 9-16 13a4 4 0 003 8c17-6 21-15 21-21 1-5-2-10-6-14z' fill='%23F2F8F6'/%3E%3Cpath d='M119 129a9 9 0 1019 0 9 9 0 00-19 0' fill='%23F1F8F3'/%3E%3C/svg%3E"
```

**Review by felixfbecker — APPROVED** [review]
Few more suggestions, otherwise LGTM!

**Pull request merged**